dsl/flatten.cc

```diff
@@ -0,0 +1,352 @@
+            }
+        }
+
+        // this moves an expression to the already-allocated space
+        void addExpr(MethodData md, unique_ptr<ast::Expression> expr) {
+            ENFORCE(md.targetLocation);
+            int idx = *md.targetLocation;
+            ENFORCE(moveQueue.size() > idx);
+            ENFORCE(moveQueue[idx].expr == nullptr);
+            moveQueue[idx] = {move(expr), md.staticLevel};
+        }
+    };
+
+    // each entry on this corresponds to a class or module scope in which we might in turn have nested methods. We only
+    // care about the innermost class scope at a given time, but the outer class may still have definitions which we
+    // need to move to the end, so we keep that below on the stack.
+    vector<ClassScope> classScopes;
+
+    int computeStaticLevel(const ast::MethodDef &methodDef) {
+        auto &methods = curMethodSet();
+        int prevLevel = methods.stack.empty() ? 0 : methods.stack.back().staticLevel;
+        return prevLevel + (methodDef.isSelf() ? 1 : 0);
+    }
+
+    void newMethodSet() {
+        classScopes.emplace_back();
+    }
+    ClassScope &curMethodSet() {
+        ENFORCE(!classScopes.empty());
+        return classScopes.back();
+    }
+    void popCurMethodSet() {
+        ENFORCE(!classScopes.empty());
+        classScopes.pop_back();
+    }
+
+    // grab all the final moved items once we're done with everything: this will grab anything that needs to be moved at
+    // the top level
+    vector<MovedItem> popCurMethodDefs() {
+        auto ret = std::move(curMethodSet().moveQueue);
+        ENFORCE(curMethodSet().stack.empty());
+        popCurMethodSet();
+        return ret;
+    };
+
+    // extract all the methods from the current queue and put them at the end of the class's current body
+    ast::ClassDef::RHS_store addClassDefMethods(core::Context ctx, ast::ClassDef::RHS_store rhs, core::Loc loc) {
+        if (curMethodSet().moveQueue.size() == 1 && rhs.size() == 1 && ast::isa_tree<ast::EmptyTree>(rhs[0].get())) {
+            // It was only 1 method to begin with, put it back
+            rhs.pop_back();
+            rhs.emplace_back(std::move(popCurMethodDefs()[0].expr));
+            return rhs;
+        }
+
+        auto exprs = popCurMethodDefs();
+        // TODO: remove all this
+        //
+        // this add the nested methods at the appropriate 'staticness level'
+
+        int highestLevel = 0;
+        for (int i = 0; i < exprs.size(); i++) {
+            auto &expr = exprs[i];
+            if (highestLevel < expr.staticLevel) {
+                highestLevel = expr.staticLevel;
+            }
+            // we need to make sure that we keep sends with their attached methods, so fix that up here
+            if (i > 0) {
+                auto send = ast::cast_tree<ast::Send>(exprs[i - 1].expr.get());
+                if (send != nullptr && send->fun == core::Names::sig()) {
+                    exprs[i - 1].staticLevel = expr.staticLevel;
```

Comment:
why is it ever mutated?
Why do we initially set it to wrong value?

Reply:
Because in the case of
```ruby
sig { ... }
def (self.)foo; end
```
we don't know ahead of time whether `foo` is going to have a `self` or not, which means we don't know the 'static level' of the `sig` until we've looked at the following method.

It's also the case that this is only important for our current broken semantics: the reason that we need to know the specific 'static level' of a `sig` block is to make sure we keep it paired with the appropriate method when we nest it into a `class << self` block. After this PR has gone in, I'll put in another PR that changes the semantics of nested `def self.f` definitions to match what Ruby actually does, in which case we will not need to generate these extra `class << self` blocks. Once we do that, we just need to always move the expressions in the order that we see them and it'll keep the `sig`s next to the methods by default, and this entire chunk of code is going to get a _lot_ simpler.